v1/plugins/logs/eventBuffer.go

```diff
@@ -0,0 +1,250 @@
+package logs
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/open-policy-agent/opa/v1/plugins/rest"
+)
+
+// eventBuffer stores and uploads a gzip compressed JSON array of EventV1 entries
+type eventBuffer struct {
+	buffer               chan []byte    // buffer stores JSON encoded EventV1 data
+	mtx                  sync.RWMutex   // mtx controls access to the buffer
+	payload              *payloadBuffer // payload contains the compressed JSON array to be uploaded
+	client               rest.Client    // client is used to upload the data to the configured service
+	uploadPath           string         // uploadPath is the configured HTTP resource path for upload
+	uploadSizeLimitBytes int64          // uploadSizeLimitBytes will enforce a maximum payload size to be uploaded
+}
+
+func newEventBuffer(bufferSizeLimitEvents int64, client rest.Client, uploadPath string, uploadSizeLimitBytes int64) *eventBuffer {
+	return &eventBuffer{
+		buffer:               make(chan []byte, bufferSizeLimitEvents),
+		payload:              newPayloadBuffer(),
+		client:               client,
+		uploadPath:           uploadPath,
+		uploadSizeLimitBytes: uploadSizeLimitBytes,
+	}
+}
+
+// Reconfigure updates the user configurable values
+func (b *eventBuffer) Reconfigure(bufferSizeLimitEvents int64, client rest.Client, uploadPath string, uploadSizeLimitBytes int64) []error {
+	b.mtx.Lock()
+	defer b.mtx.Unlock()
+
+	b.client = client
+	b.uploadPath = uploadPath
+	oldUploadSizeLimitBytes := b.uploadSizeLimitBytes
+	b.uploadSizeLimitBytes = uploadSizeLimitBytes
+
+	if int64(cap(b.buffer)) == bufferSizeLimitEvents && uploadSizeLimitBytes == oldUploadSizeLimitBytes {
+		return nil
+	}
+
+	close(b.buffer)
+	newBuffer := make(chan []byte, bufferSizeLimitEvents)
+
+	var errs []error
+	for event := range b.buffer {
+		if int64(len(event)) < b.uploadSizeLimitBytes {
+			push(newBuffer, event)
+			continue
+		}
+
+		var decoded EventV1
+		if err := json.Unmarshal(event, &decoded); err != nil {
+			continue
+		}
+		if err := b.dropNDCache(&decoded, event); err != nil {
+			if !errors.Is(err, droppedNDCache{}) {
+				errs = append(errs, err)
+				continue
+			}
+			event, err = json.Marshal(decoded)
+			if err != nil {
+				errs = append(errs, err)
+				continue
+			}
+		}
+		push(newBuffer, event)
+	}
+	b.buffer = newBuffer
+
+	return errs
+}
+
+// Push attempts to add a new event to the buffer
+func (b *eventBuffer) Push(event EventV1) error {
+	encoded, err := json.Marshal(event)
+	if err != nil {
+		return err
+	}
+
+	err = b.dropNDCache(&event, encoded)
+	if err != nil {
+		if !errors.As(err, &droppedNDCache{}) {
+			return err
+		}
+	}
+
+	// This only blocks when the buffer is being reconfigured
+	b.mtx.RLock()
+	defer b.mtx.RUnlock()
+
+	push(b.buffer, encoded)
+
+	return err
```

Comment:
💭 I've been wondering about this err, i.e. "why not return nil?" -- but it could be a non-nil error from line 88, dropNDCache. This makes me wonder if push() is OK to call if dropNDCache returned an error -- Is it?

Reply:
Yes calling push is ok, the dropNDCache error is only used to communicate that the ND cache was dropped so that the metric can be incremented and the message logged. If the event meets the size limit requirement after dropping the ND cache, then it is ok to be pushed into the buffer.